### PR TITLE
Addition of the 'Variable Keyhole' sector to the set of AAT turnpoint types. (Ticket 1687)

### PR DESCRIPTION
**Pull request by Exadios**
Re submission of Ticket 1687

### CODE REVIEW

**Comment by lordfolken** [comment]
Please change the title of this to something meaningful @Exadios .

**Comment by Exadios** [comment]
Could someone review this PR please?

**Comment by MaxKellermann** [comment]
The first commit looks like work in progress, but work in progress is not eligible for merging.
The second commit appears to redo the first one; should the two be folded?
The second commit message is confusing.

File: .gitignore [hunk 1]
```diff
@@ -32,3 +32,4 @@ build/local-config.mk
 *.pyc
 *.pyo
 *.bak
+.*.swp
```
Comment (MaxKellermann) [inline .gitignore]: Why?

File: build/libtask.mk [hunk 2]
```diff
@@ -52,6 +52,7 @@ TASK_SOURCES = \
 	$(TASK_SRC_DIR)/ObservationZones/SymmetricSectorZone.cpp \
 	$(TASK_SRC_DIR)/ObservationZones/KeyholeZone.cpp \
 	$(TASK_SRC_DIR)/ObservationZones/AnnularSectorZone.cpp \
+  $(TASK_SRC_DIR)/ObservationZones/VariableKeyholeZone.cpp \
```
Comment (MaxKellermann) [inline build/libtask.mk]: Bad indent

File: build/main.mk [hunk 3]
```diff
@@ -115,6 +115,7 @@ DIALOG_SOURCES = \
 	$(SRC)/Dialogs/Task/Widgets/LineSectorZoneEditWidget.cpp \
 	$(SRC)/Dialogs/Task/Widgets/SectorZoneEditWidget.cpp \
 	$(SRC)/Dialogs/Task/Widgets/KeyholeZoneEditWidget.cpp \
+  $(SRC)/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp \
```
Comment (MaxKellermann) [inline build/main.mk]: bad indent

File: build/test.mk [hunk 4]
```diff
@@ -787,7 +787,8 @@ DEBUG_PROGRAM_NAMES = \
 	IGC2NMEA \
 	NearestWaypoints \
 	RunKalmanFilter1d \
-	ArcApprox
+	ArcApprox \
+  VariableKeyholeZone
```
Comment (MaxKellermann) [inline build/test.mk]: Move this one line up to avoid moving the backslash around

File: doc/XCSoar.doxyfile [hunk 5]
```diff
@@ -474,7 +474,7 @@ SHOW_USED_FILES = YES
 # then setting the SHOW_DIRECTORIES tag to YES will show the directory hierarchy
 # in the documentation. The default is NO.
 
-SHOW_DIRECTORIES = NO
+SHOW_DIRECTORIES = YES
```
Comment (MaxKellermann) [inline doc/XCSoar.doxyfile]: Why?

File: doc/XCSoar.doxyfile [hunk 6]
```diff
@@ -739,7 +739,7 @@ VERBATIM_HEADERS = YES
 # of all compounds will be generated. Enable this if the project
 # contains a lot of classes, structs, unions or interfaces.
 
-ALPHABETICAL_INDEX = NO
+ALPHABETICAL_INDEX = YES
```
Comment (MaxKellermann) [inline doc/XCSoar.doxyfile]: Why?

File: doc/XCSoar.doxyfile [hunk 7]
```diff
@@ -1435,7 +1435,7 @@ CALL_GRAPH = YES
 # the time of a run. So in most cases it will be better to enable caller
 # graphs for selected functions only using the \callergraph command.
 
-CALLER_GRAPH = NO
+CALLER_GRAPH = YES
```
Comment (MaxKellermann) [inline doc/XCSoar.doxyfile]: Why?

File: build/test.mk [hunk 8]
```diff
@@ -962,6 +963,16 @@ DOWNLOAD_FILE_SOURCES = \
 DOWNLOAD_FILE_DEPENDS = LIBNET IO THREAD UTIL
 $(eval $(call link-program,DownloadFile,DOWNLOAD_FILE))
 
+CANVAS_KEYHOLE_SOURCES = \
+  $(TEST_SRC_DIR)/CanvasKeyhole.cpp
```
Comment (MaxKellermann) [inline build/test.mk]: File doesn't exist

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "VariableKeyholeZoneEditWidget.hpp"
+#include "Engine/Task/ObservationZones/KeyholeZone.hpp"
+#include "Language/Language.hpp"
+
+enum Controls
+  {
```
Comment (MaxKellermann) [inline src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp]: Bad indent

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+//------------------------------------------------------------------------------
+VariableKeyholeZoneEditWidget::VariableKeyholeZoneEditWidget(VariableKeyholeZone &_oz)
+  :ObservationZoneEditWidget(_oz) {}
+
+//------------------------------------------------------------------------------
+void
+VariableKeyholeZoneEditWidget::Prepare(ContainerWindow &parent,
+                                         const PixelRect &rc)
+  {
```
Comment (MaxKellermann) [inline src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp]: Bad indent

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+                 this->GetObject().GetEndRadial(),
+                 10,
+                 true,
+                 this);
+  }
+
+//------------------------------------------------------------------------------
+bool
+VariableKeyholeZoneEditWidget::Save(bool &_changed)
+  {
```
Comment (MaxKellermann) [inline src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp]: Bad indent

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.hpp [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+#include "ObservationZoneEditWidget.hpp"
+#include "Engine/Task/ObservationZones/VariableKeyholeZone.hpp"
+
+/**
+ * The purpose of this class is to enter or edit the so called
+ * 'VariableKeyholeZone' observation zone type.
+ */
+class VariableKeyholeZoneEditWidget : public ObservationZoneEditWidget
+  {
```
Comment (MaxKellermann) [inline src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.hpp]: Bad indent

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.hpp [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  VariableKeyholeZoneEditWidget(VariableKeyholeZone &_oz);
+
+protected:
+  /**
+   * The const version of the observation zone.
+   * @return The OZ for const operations.
+   */
+  const VariableKeyholeZone &GetObject() const
+    {
```
Comment (MaxKellermann) [inline src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.hpp]: Bad indent

File: src/Dialogs/Task/dlgTaskHelpers.cpp [hunk 14]
```diff
@@ -188,6 +189,14 @@ OrderedTaskPointRadiusLabel(const ObservationZonePoint &ozp, TCHAR* buffer)
                        Units::GetDistanceName());
     return;
 
+  case ObservationZone::Shape::VARIABLE_KEYHOLE:
+    StringFormatUnsafe(buffer, _T("%s - %s: %.1f%s"),
+                               _("Variable Sector"),
```
Comment (MaxKellermann) [inline src/Dialogs/Task/dlgTaskHelpers.cpp]: Bad indent

File: src/Engine/Task/Factory/AATTaskFactory.cpp [hunk 15]
```diff
@@ -86,7 +87,7 @@ AATTaskFactory::GetMutatedPointType(const OrderedTaskPoint &tp) const
   case TaskPointFactoryType::FAI_SECTOR:
   case TaskPointFactoryType::SYMMETRIC_QUADRANT:
     newtype = TaskPointFactoryType::AAT_CYLINDER;
-    //ToDo: create a 90 degree symmetric AAT sector
+    // TODO: create a 90 degree symmetric AAT sector
```
Comment (MaxKellermann) [inline src/Engine/Task/Factory/AATTaskFactory.cpp]: Why?

File: src/Engine/Task/Factory/AbstractTaskFactory.cpp [hunk 16]
```diff
@@ -162,120 +166,142 @@ AbstractTaskFactory::CreateFinish(WaypointPtr wp) const
...
-    case ObservationZone::Shape::BGAFIXEDCOURSE:
-    case ObservationZone::Shape::BGAENHANCEDOPTION:
-    case ObservationZone::Shape::ANNULAR_SECTOR:
-      return TaskPointFactoryType::START_CYLINDER;
-
-    case ObservationZone::Shape::BGA_START:
-      return TaskPointFactoryType::START_BGA;
-    }
-    break;
+  switch (point.GetType())
```
Comment (MaxKellermann) [inline src/Engine/Task/Factory/AbstractTaskFactory.cpp]: Why did you reformat this whole section? (badly)

File: src/Engine/Task/Factory/AbstractTaskFactory.cpp [hunk 17]
```diff
@@ -162,120 +166,142 @@ AbstractTaskFactory::CreateFinish(WaypointPtr wp) const
   return CreateFinish(GetDefaultFinishType(), std::move(wp));
 }
 
+//------------------------------------------------------------------------------
 TaskPointFactoryType 
 AbstractTaskFactory::GetType(const OrderedTaskPoint &point) const
-{
+  {
```
Comment (MaxKellermann) [inline src/Engine/Task/Factory/AbstractTaskFactory.cpp]: Why??

File: src/Engine/Task/Factory/AbstractTaskFactory.cpp [hunk 18]
```diff
@@ -162,120 +166,142 @@ AbstractTaskFactory::CreateFinish(WaypointPtr wp) const
   return CreateFinish(GetDefaultFinishType(), std::move(wp));
 }
 
+//------------------------------------------------------------------------------
```
Comment (MaxKellermann) [inline src/Engine/Task/Factory/AbstractTaskFactory.cpp]: Why?

File: src/Engine/Task/ObservationZones/Boundary.hpp [hunk 19]
```diff
@@ -33,6 +33,13 @@ class OZBoundary : public std::forward_list<GeoPoint> {
   /**
    * Generate boundary points for the arc described by the parameters.
    * This excludes the points at the start/end angle.
+   *
+   * The arc is generated in an increasing angle order - i.e. CW.
```
Comment (MaxKellermann) [inline src/Engine/Task/ObservationZones/Boundary.hpp]: Adding more documentation is good, but not hidden inside this unrelated patch.

File: src/Engine/Task/ObservationZones/VariableKeyholeZone.hpp [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+   * @param can_start_through_top When used as a start zone a start can be
+   *                              triggered by crossing the "top". (What is
+   *                              the top) (Check callers to
+   *                              ObservationZone::CanStartThroughTop()).
+   * @param arc_boundary Does the boundary include the arc? I.e. is crossing
+   *                     the arc scored?
+   * @param loc The location at the center of the inner radius.
+   * @param radiusOuter The location of the outer sector.
+   * @param startRadial The first radial of the sector in a CW sense, north
+                        zero.
```
Comment (MaxKellermann) [inline src/Engine/Task/ObservationZones/VariableKeyholeZone.hpp]: Missing `*`

File: src/Engine/Task/ObservationZones/VariableKeyholeZone.hpp [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+ * equal to, that of the outer radius.
+ *
+ * By manipulation of the inner, outer radii, and the start and end radial
+ * this class can handle so called sector, arbitrary keyhole and circle
+ * observation zones.
+ */
+
+class VariableKeyholeZone final : public SectorZone
+{
+private:
```
Comment (MaxKellermann) [inline src/Engine/Task/ObservationZones/VariableKeyholeZone.hpp]: Redundant `private`

File: src/Renderer/OZRenderer.cpp [hunk 22]
```diff
@@ -63,7 +64,7 @@ OZRenderer::Prepare(Canvas &canvas, Layer layer, int offset) const
 #endif /* !GDI */
 
     canvas.SelectNullPen();
-    
+
```
Comment (MaxKellermann) [inline src/Renderer/OZRenderer.cpp]: Why?

File: src/Renderer/OZRenderer.cpp [hunk 23]
```diff
@@ -127,7 +128,7 @@ OZRenderer::Draw(Canvas &canvas, Layer layer, const Projection &projection,
     if (layer != LAYER_INACTIVE) {
       auto p_center = projection.GeoToScreen(oz.GetReference());
       canvas.DrawCircle(p_center.x, p_center.y,
-                    projection.GeoToScreenDistance(oz.GetRadius()));
+                        projection.GeoToScreenDistance(oz.GetRadius()));
```
Comment (MaxKellermann) [inline src/Renderer/OZRenderer.cpp]: Why?

File: src/Task/Deserialiser.cpp [hunk 24]
```diff
@@ -108,18 +110,24 @@ DeserialiseOZ(const Waypoint &wp, const ConstDataNode &node, bool is_turnpoint)
       ls->SetLength(length);
 
     return ls;
-  } else if (StringIsEqual(type, _T("Cylinder"))) {
+  }
+
+  else if (StringIsEqual(type, _T("Cylinder"))) {
```
Comment (MaxKellermann) [inline src/Task/Deserialiser.cpp]: Why??

File: src/Task/Deserialiser.cpp [hunk 25]
```diff
@@ -93,6 +94,7 @@ DeserialiseWaypoint(const ConstDataNode &node, const Waypoints *waypoints)
   return WaypointPtr(wp);
 }
 
+//------------------------------------------------------------------------------
```
Comment (MaxKellermann) [inline src/Task/Deserialiser.cpp]: Why?

File: src/Task/Deserialiser.cpp [hunk 26]
```diff
@@ -108,18 +110,24 @@ DeserialiseOZ(const Waypoint &wp, const ConstDataNode &node, bool is_turnpoint)
...
 
     double radius;
     if (node.GetAttribute(_T("radius"), radius) && radius > 0)
       ls->SetRadius(radius);
 
     return ls;
-  } else if (StringIsEqual(type, _T("MatCylinder"))) {
+  }
+
+  else if (StringIsEqual(type, _T("MatCylinder"))) {
```
Comment (MaxKellermann) [inline src/Task/Deserialiser.cpp]: Why???

File: src/Task/Deserialiser.cpp [hunk 27]
```diff
@@ -108,18 +110,24 @@ DeserialiseOZ(const Waypoint &wp, const ConstDataNode &node, bool is_turnpoint)
...
     return CylinderZone::CreateMatCylinderZone(wp.location);
-  } else if (StringIsEqual(type, _T("Sector"))) {
+  }
 
+  else if (StringIsEqual(type, _T("Sector")))
+    {
```
Comment (MaxKellermann) [inline src/Task/Deserialiser.cpp]: Seriously, why???

File: src/Task/Deserialiser.cpp [hunk 28]
```diff
@@ -146,9 +154,31 @@ DeserialiseOZ(const Waypoint &wp, const ConstDataNode &node, bool is_turnpoint)
     node.GetAttribute(_T("radius"), radius);
 
     return new SymmetricSectorZone(wp.location, radius);
-  } else if (StringIsEqual(type, _T("Keyhole")))
+  }
+
+  else if (StringIsEqual(type, _T("Keyhole")))
```
Comment (MaxKellermann) [inline src/Task/Deserialiser.cpp]: WTF!

File: src/Task/Deserialiser.cpp [hunk 29]
```diff
@@ -160,10 +190,14 @@ DeserialiseOZ(const Waypoint &wp, const ConstDataNode &node, bool is_turnpoint)
       KeyholeZone::CreateCustomKeyholeZone(wp.location, radius, angle);
     keyhole->SetInnerRadius(inner_radius);
     return keyhole;
-  } else if (StringIsEqual(type, _T("BGAStartSector")))
+  }
+
+  else if (StringIsEqual(type, _T("BGAStartSector")))
```
Comment (MaxKellermann) [inline src/Task/Deserialiser.cpp]: WTF ...

File: src/Task/Deserialiser.cpp [hunk 30]
```diff
@@ -160,10 +190,14 @@ DeserialiseOZ(const Waypoint &wp, const ConstDataNode &node, bool is_turnpoint)
...
     return KeyholeZone::CreateBGAStartSectorZone(wp.location);
+
```
Comment (MaxKellermann) [inline src/Task/Deserialiser.cpp]: Why?

File: src/Task/Deserialiser.cpp [hunk 31]
```diff
@@ -160,10 +190,14 @@ DeserialiseOZ(const Waypoint &wp, const ConstDataNode &node, bool is_turnpoint)
...
   else if (StringIsEqual(type, _T("BGAFixedCourse")))
     return KeyholeZone::CreateBGAFixedCourseZone(wp.location);
+
```
Comment (MaxKellermann) [inline src/Task/Deserialiser.cpp]: Why?

File: src/Task/Serialiser.cpp [hunk 32]
```diff
@@ -111,6 +111,17 @@ Visit(WritableDataNode &node, const AnnularSectorZone &data)
   node.SetAttribute(_T("inner_radius"), data.GetInnerRadius());
 }
 
+//------------------------------------------------------------------------------
```
Comment (MaxKellermann) [inline src/Task/Serialiser.cpp]: What use is this ASCII art line?

File: src/Task/Serialiser.cpp [hunk 33]
```diff
@@ -28,7 +28,7 @@
 #include "Task/ObservationZones/LineSectorZone.hpp"
 #include "Task/ObservationZones/KeyholeZone.hpp"
 #include "Task/ObservationZones/AnnularSectorZone.hpp"
-#include "Task/ObservationZones/SymmetricSectorZone.hpp"
```
Comment (MaxKellermann) [inline src/Task/Serialiser.cpp]: Why remove this line?

File: src/Task/TaskFileSeeYou.cpp [hunk 34]
```diff
@@ -189,11 +190,15 @@ ParseOZs(SeeYouTurnpointInformation turnpoint_infos[], const TCHAR *params[],
     return -1;
 
   turnpoint_infos[oz_index].valid = true;
+  // Defaults
```
Comment (MaxKellermann) [inline src/Task/TaskFileSeeYou.cpp]: What use is this comment? I don't understand what it means.

File: src/Task/TaskFileSeeYou.cpp [hunk 35]
```diff
@@ -189,11 +190,15 @@ ParseOZs(SeeYouTurnpointInformation turnpoint_infos[], const TCHAR *params[],
...
+  turnpoint_infos[oz_index].radius2 = 0.0;
+
   // Iterate through available OZ options
   for (unsigned i = 1; i < n_params; i++) {
     const TCHAR *pair = params[i];
     SeeYouTurnpointInformation &tp_info = turnpoint_infos[oz_index];
 
+
```
Comment (MaxKellermann) [inline src/Task/TaskFileSeeYou.cpp]: Why?

File: src/Task/TaskFileSeeYou.cpp [hunk 36]
```diff
@@ -292,9 +297,9 @@ gcc_pure
 static Angle
 CalcIntermediateAngle(const SeeYouTurnpointInformation &turnpoint_infos,
                       const GeoPoint &location,
-                      const GeoPoint &start,
```
Comment (MaxKellermann) [inline src/Task/TaskFileSeeYou.cpp]: Why reorder parameters?

File: src/Task/TaskFileSeeYou.cpp [hunk 37]
```diff
@@ -385,15 +390,20 @@ CreateOZ(const SeeYouTurnpointInformation &turnpoint_infos,
     const Angle RadialStart = (A12adj - turnpoint_infos.angle1).AsBearing();
     const Angle RadialEnd = (A12adj + turnpoint_infos.angle1).AsBearing();
 
-    if (turnpoint_infos.radius2 > 0 &&
-        (turnpoint_infos.angle2.AsBearing().Degrees()) < 1) {
-      oz = new AnnularSectorZone(wp->location, turnpoint_infos.radius1,
-          RadialStart, RadialEnd, turnpoint_infos.radius2);
+    if (turnpoint_infos.radius2 > 0.0 &&
```
Comment (MaxKellermann) [inline src/Task/TaskFileSeeYou.cpp]: Why change `0` to `0.0`?

File: src/Task/TaskFileSeeYou.cpp [hunk 38]
```diff
@@ -385,15 +390,20 @@ CreateOZ(const SeeYouTurnpointInformation &turnpoint_infos,
...
     } else {
-      oz = new SectorZone(wp->location, turnpoint_infos.radius1,
-          RadialStart, RadialEnd);
+      oz = new AnnularSectorZone(wp->location,
+                                 turnpoint_infos.radius1,
+                                 RadialStart,
+                                 RadialEnd,
+                                 turnpoint_infos.radius2);
     }
-
```
Comment (MaxKellermann) [inline src/Task/TaskFileSeeYou.cpp]: Why?

File: src/Task/TaskFileSeeYou.hpp [hunk 39]
```diff
@@ -26,11 +26,22 @@
...
+ * A class that reads and parses a SeeYou task file to an XCSoar internal
+ * task representation.
+ */
+
 class TaskFileSeeYou: public TaskFile
 {
 public:
-  explicit TaskFileSeeYou(Path _path):TaskFile(_path) {}
-
+  explicit TaskFileSeeYou(Path path):TaskFile(path) {}
```
Comment (MaxKellermann) [inline src/Task/TaskFileSeeYou.hpp]: Why?

File: src/Task/TaskFileSeeYou.hpp [hunk 40]
```diff
@@ -26,11 +26,22 @@
 #include "TaskFile.hpp"
 #include "OS/Path.hpp"
 
+/**
+ * A class that reads and parses a SeeYou task file to an XCSoar internal
```
Comment (MaxKellermann) [inline src/Task/TaskFileSeeYou.hpp]: Documentation belongs in a separate commit because it's unrelated to this new feature.

File: src/Task/TaskFileSeeYou.hpp [hunk 41]
```diff
@@ -26,11 +26,22 @@
...
+
 class TaskFileSeeYou: public TaskFile
 {
 public:
-  explicit TaskFileSeeYou(Path _path):TaskFile(_path) {}
-
+  explicit TaskFileSeeYou(Path path):TaskFile(path) {}
+
+  /**
+   * Give the task produced by parsing the SeeYou file.
```
Comment (MaxKellermann) [inline src/Task/TaskFileSeeYou.hpp]: Documentation belongs in a separate commit because it's unrelated to this new feature.

File: test/src/VariableKeyholeZone.cpp [hunk 42]
```diff
@@ -0,0 +1,125 @@
...
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#define USE_FREETYPE 1
```
Comment (MaxKellermann) [inline test/src/VariableKeyholeZone.cpp]: WTF??

File: test/src/VariableKeyholeZone.cpp [hunk 43]
```diff
@@ -0,0 +1,125 @@
...
+#define ENABLE_OPENGL 1
```
Comment (MaxKellermann) [inline test/src/VariableKeyholeZone.cpp]: WTF???

File: test/src/VariableKeyholeZone.cpp [hunk 44]
```diff
@@ -0,0 +1,125 @@
...
+#include "Math/Angle.hpp"
+#include "Math/FastTrig.hpp"
+
+#include <string.h>
+#include <stdlib.h>
+#include <iostream>
+#include <fstream>
+
+struct Args
+  {
```
Comment (MaxKellermann) [inline test/src/VariableKeyholeZone.cpp]: Bad indent

File: test/src/VariableKeyholeZone.cpp [hunk 45]
```diff
@@ -0,0 +1,125 @@
...
+  GeoPoint ref;
+  double outer_radius;
+  double inner_radius;
+  Angle start_radial;
+  Angle end_radial;
+
+  Args(int argc, const char *argv[])
+    : gnuplot_file(""),
+      ref(Angle::Zero(), Angle::Zero()),
+      outer_radius(10000),
```
Comment (MaxKellermann) [inline test/src/VariableKeyholeZone.cpp]: Use C++11 initializers

File: test/src/VariableKeyholeZone.cpp [hunk 46]
```diff
@@ -0,0 +1,125 @@
...
+  std::ofstream gnuplot;
+
+  GeoPoint ref;
+  double outer_radius;
+  double inner_radius;
+  Angle start_radial;
+  Angle end_radial;
+
+  Args(int argc, const char *argv[])
+    : gnuplot_file(""),
```
Comment (MaxKellermann) [inline test/src/VariableKeyholeZone.cpp]: What's the point of this initializer?

File: test/src/VariableKeyholeZone.cpp [hunk 47]
```diff
@@ -0,0 +1,125 @@
...
+          {
+          i++;
+          this->ref.latitude = Angle::Degrees(atof(argv[i]));
+          }
+        }
+      }
+    }
+  };
+
+void Test(Args &);
```
Comment (MaxKellermann) [inline test/src/VariableKeyholeZone.cpp]: What's the point of this declaration?

File: test/src/harness_task.cpp [hunk 48]
```diff
@@ -66,6 +67,11 @@ class ObservationZoneVisitorPrint
     printf("# symmetric quadrant\n");
   }
 
+  void Visit(const VariableKeyholeZone &oz)
+    {
```
Comment (MaxKellermann) [inline test/src/harness_task.cpp]: Bad indent

File: src/Engine/Task/ObservationZones/Boundary.hpp (hunk 19 above)
Comment (Exadios) [reply src/Engine/Task/ObservationZones/Boundary.hpp]: So, just a question on the logistics:  For the purposes of the PR should this commit be on a branch of its own or should it be broken out as a separate on this branch - i.e. 'ticket_1687_on_master'?

Comment (MaxKellermann) [reply src/Engine/Task/ObservationZones/Boundary.hpp]: My stance: can be same branch in same PR. As long as it's a separate commit, I can easily cherry-pick good commits and postpone the others.

File: test/src/VariableKeyholeZone.cpp (hunk 47 above)
Comment (Exadios) [reply test/src/VariableKeyholeZone.cpp]: -Werror=missing-declarations

File: src/Renderer/OZRenderer.cpp (hunk 22 above)
Comment (Exadios) [reply src/Renderer/OZRenderer.cpp]: Extra spaces.

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp (hunk 9 above)
Comment (Exadios) [reply src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp]: Whitesmith.

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp (hunk 10 above)
Comment (Exadios) [reply src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp]: Whitesmith.

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp (hunk 11 above)
Comment (Exadios) [reply src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.cpp]: Whitesmith.

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.hpp (hunk 12 above)
Comment (Exadios) [reply src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.hpp]: Whitesmith.

File: src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.hpp (hunk 13 above)
Comment (Exadios) [reply src/Dialogs/Task/Widgets/VariableKeyholeZoneEditWidget.hpp]: Whitesmith.

File: test/src/VariableKeyholeZone.cpp (hunk 44 above)
Comment (Exadios) [reply test/src/VariableKeyholeZone.cpp]: Whitesmith.

File: test/src/VariableKeyholeZone.cpp (hunk 47 above)
Comment (MaxKellermann) [reply test/src/VariableKeyholeZone.cpp]: This is a lousy workaround for a function you forgot to make `static`.

File: src/Engine/Task/Factory/AATTaskFactory.cpp [hunk 49]
```diff
@@ -45,6 +45,7 @@ static constexpr LegalPointSet aat_im_types{
   TaskPointFactoryType::AAT_SEGMENT,
   TaskPointFactoryType::AAT_ANNULAR_SECTOR,
   TaskPointFactoryType::AAT_KEYHOLE,
+  TaskPointFactoryType::AAT_VARIABLE_KEYHOLE_SECTOR,
```
Comment (MaxKellermann) [inline src/Engine/Task/Factory/AATTaskFactory.cpp]:
```
src/Engine/Task/Factory/AATTaskFactory.cpp:48:25: error: no member named 'AAT_VARIABLE_KEYHOLE_SECTOR' in 'TaskPointFactoryType'
```

**Comment by MaxKellermann** [comment]
I've cherry-picked the documentation commits; please rebase and clean up the remaining commits for review.

File: src/Engine/Task/Factory/AATTaskFactory.cpp (hunk 49 above)
Comment (Exadios) [reply src/Engine/Task/Factory/AATTaskFactory.cpp]: Huh? Are you sure you had the working directory up to date with the HEAD?

**Comment by Exadios** [comment]
> I've cherry-picked the documentation commits; please rebase and clean up the remaining commits for review.

Do you want me to remove the cherry picked commit from the branch?

**Comment by MaxKellermann** [comment]
> Do you want me to remove the cherry picked commit from the branch?

I asked you to rebase. Rebasing will drop merged commits automatically.

Closing this PR because there was no progress for more than 4 months.

**Comment by kerel-fs** [comment]
@Exadios I rebased your patches & did a clean-up of the two remaining commits (squashed them and reworded the resulting commit message). The result of this can be found in https://gitlab.com/kerel-fs/xcsoar/tree/ticket1687

If you are stilll insterested in this PR please review my changes and push these or an otherwise updated patch-set to this branch and re-open this PR.

**Comment by Exadios** [comment]
@kerel-fs Thanks. That's mighty decent of you. I had actually forgotten about this pull request.

I have pulled your commit. I'll have a look and do some quick tests. But that will probably have to wait until I get down to the club - in a couple of weeks. (For one thing I have to update my build machine.)

I'll get back to you.
